File: algorithms/1_easy/D/03069_distribute_elements_into_two_arrays_i/two_pointers.cpp
class Solution {
public:
    static vector<int> resultArray(const vector<int>& nums) {
        if (nums.size() <= 2) { return nums; }

        const size_t len = nums.size();
        auto res = vector<int>(len);
        size_t hi = len - 1;
        res[hi] = nums[1];
        size_t lo = 0;
        res[lo] = nums[0];

        for (size_t i = 2; i < len; i += 1) {
            if (res[hi] > res[lo]) {
                hi -= 1;
                res[hi] = nums[i];
            } else {
                lo += 1;
                res[lo] = nums[i];
            }
        }

        std::reverse(res.begin() + hi, res.end());

        return res;
    }
};

File: algorithms/1_easy/D/03069_distribute_elements_into_two_arrays_i/two_pointers.rs
impl Solution {
    pub fn result_array(nums: Vec<i32>) -> Vec<i32> {
        if nums.len() <= 2 { return nums; }

        let len = nums.len();
        let mut res = vec![0_i32; len];
        let mut hi = len - 1;
        res[hi] = nums[1];
        let mut lo = 0;
        res[lo] = nums[0];

        for i in 2..len {
            if res[hi] > res[lo] {
                hi -= 1;
                res[hi] = nums[i];
            } else {
                lo += 1;
                res[lo] = nums[i];
            }
        }

        res[hi..len].reverse();

        res
    }
}
